### PR TITLE
Introducing histogram* bingroup attributes

### PR DESCRIPTION
**Pull request by etpinard**
resolves https://github.com/plotly/plotly.js/issues/3749 (the 1d overlay case) and resolves https://github.com/plotly/plotly.js/issues/3515 (the 2d case).

Users can now group traces such that their autobin settings are _compatible_ (see https://github.com/plotly/plotly.js/issues/3515#issuecomment-460706202 for more info on what _compatible_ means).

Examples:
- https://codepen.io/etpinard/pen/YbyOKO?editors=0010 (1d overlay)
- https://codepen.io/etpinard/pen/arvazV (2d)

In brief, this PR:

- adds attribute `bingroup` to `histogram`, `histogram2d` and `histogram2dcontour` traces
- `histogram2d` and `histogram2dcontour` traces also get new attributes `xbingroup` and `ybingroup` to allow users to group x-bins and y-bins separately.

To do so, I decided to:

- make all histogram* traces (i.e. 1d and 2d) go through the same `crossTraceDefaults` method
- make all histogram* traces use that same `calcAllAutoBins` routine

@plotly/plotly_js best to review this PR commit-by-commit. Starting by reviewing the jasmine tests added in https://github.com/plotly/plotly.js/commit/359fbbc9267fa2f241062dc63cbadb4d1130b3d2 might make things easier to grasp.

### CODE REVIEW

File: test/jasmine/tests/histogram_test.js [hunk 1]
```diff
@@ -186,6 +186,343 @@ describe('Test histogram', function() {
...
+
+                    {uid: 'e', type: 'histogram', y: [3]},
+                    {uid: 'f', type: 'histogram', y: [2], xaxis: 'x2'},
+
+                    {uid: 'g', type: 'histogram', x: [1]},
+                    {uid: 'h', type: 'histogram', x: [2], yaxis: 'y2'},
+                    {uid: 'i', type: 'histogram', x: [2]}
+                ],
+                layout: {barmode: 'overlay'}
+            };
```
Comment (antoinerg) [inline test/jasmine/tests/histogram_test.js]: Interestingly, this mock does not render in this PR and in `master`. :thinking: 

Screenshot_2019-05-17_17-13-22

File: test/jasmine/tests/histogram_test.js [hunk 2]
```diff
@@ -186,6 +186,343 @@ describe('Test histogram', function() {
...
+                    {bingroup: '2', type: 'histogram', y: [1], xaxis: 'x2'},
+                    {bingroup: '1', type: 'histogram', y: [3], xaxis: 'x2'},
+                    {bingroup: '2', type: 'histogram', y: [3]},
+                    {uid: 'f', type: 'histogram', y: [2], xaxis: 'x2'},
+                    {bingroup: '3', type: 'histogram', x: [1]},
+                    {bingroup: '1', type: 'histogram', x: [2], yaxis: 'y2'},
+                    {bingroup: '3', type: 'histogram', x: [2]}
+                ],
+                layout: {barmode: 'overlay'}
+            };
```
Comment (antoinerg) [inline test/jasmine/tests/histogram_test.js]: This mock also doesn't render

File: test/jasmine/tests/histogram_test.js [hunk 3]
```diff
@@ -186,6 +186,343 @@ describe('Test histogram', function() {
...
+        });
+
+        it('should not group traces across different calendars', function() {
+            gd = {
+                data: [
+                    {uid: 'a', bingroup: '1', type: 'histogram', x: ['2000-01-01']},
+                    {uid: 'b', bingroup: '1', type: 'histogram', x: ['2000-01-01'], xcalendar: 'julian'},
+                ],
+                layout: {barmode: 'overlay'}
+            };
```
Comment (antoinerg) [inline test/jasmine/tests/histogram_test.js]: This mock also doesn't render

**Comment by antoinerg** [comment]
Thanks for the very nice PR @etpinard ! The tests are very nice :+1: 

I think I have uncovered a bug (probably not related to your PR) when reviewing. A few mocks in your test suite do not render! I will leave it up to you to decide whether to fix this here or in another PR. Let me know what you decide to do.

**Comment by etpinard** [comment]
> I think I have uncovered a bug (not related to your PR) when reviewing

Thanks for noticing that! Something is (still) up with single-value histograms under `barmode: overlay`. Tracked now in https://github.com/plotly/plotly.js/issues/3881. Commit 3776cd2da makes the data/layout used in the new `crossTraceDefaults` tests render ok.

**Pull request merged**